### DISCUSSION TITLE
Using molten for lua jupyter kernel - works but not aware of vim globals

### DISCUSSION THREAD

**Discussion by Or1g3n** [benlubas/molten-nvim#261] [Q&A] [answered]
Firstly, molten.nvim is an amazing plugin and it was the last piece of the puzzle to fully replace my workflow in vscode.

I know neovim has a lua interpreter built-in but I really have grown accustomed to the workflow offered by molten when developing ipynb notebooks. So I decided to pip install ilua which allowed me to run lua code inside a neovim buffer molten style. The only issue I have run into is it does not recognize the vim global in neovim which allows you to use vim api functions.

Has anyone experimented with lua kernel via molten and have you overcome this issue? The goal is for the ilua kernel to be aware of the vim api when molten initializes the kernel so I do not get "the global variable vim is a nil value" (something like that).

**Comment by benlubas** [comment]
My first thought is that you'd have to alter the ilua kernel so that it runs all of nvim's runtime files on startup.. which.. I mean. I think you would be better off implementing the juptyer kernel protocol in lua as a plugin and using the current neovim instance to execute code. That would be a large undertaking though.


But... If you're okay with the code you execute not effecting your running neovim session (which honestly might be more desirable). You might be able to use https://github.com/mfussenegger/nlua as the lua runtime. iLua claims to support even weird runtimes such as this. So I'm hopeful that it would work.

The second approach is the first I'd explore. If you get it working definitely let me know, that would be really cool!

**Reply by Or1g3n** [reply]
Really appreciate your thoughts! I'll start exploring option 2 and post back with anything interesting.

**Comment by Or1g3n** [comment]
My work computer is windows and since nlua was written using linux shebang style script, i'll have to do some gymnastics to tweak the script to remove unix specific commands and convert the file into an .exe. At any rate, my personal machine is linux so testing out ilua with nlua was very straight forward and it appears to work beautifully. Thanks again for making me aware of nlua.

Only steps were to:
```bash
pip install ilua
```
```bash
sudo luarocks install nlua
```
```bash
ilua -i nlua
```

## Result - I was able to run neovim api aware commands in ilua interactive shell with autocomplete. 
![image](https://github.com/user-attachments/assets/7f9046fe-3af7-414e-af94-aa445d8fcd6c)

## Next question
Does Molten have any type of config for the kernels that it calls. So rather than initializing ilua, I could tell it to call  ```ilua -i nlua``` instead when starting the kernel?

**Reply by benlubas** [reply]
wow that was super simple!

Molten works with jupyter kernels, which are defined by kernel spec files (by default these files are located in `~/.local/share/jupyter/kernels`)

I think ilua should be able to create one of those files, but I just installed it and read the help and it doesn't seem like it can. so that's... interesting. It's also unable to run in "server mode" so it seems like you have to run the python applications internals inorder to create a jupyter kernel.

I found this example (from [here](https://github.com/guysv/ilua/issues/15)) (I've switched to use nlua):

```json
{
	"argv":
	[
		"python",
		"-m",
		"ilua.app",
		"-c",
		"{connection_file}",
		"-i",
		"nlua"
	],
	"display_name": "Nvim Lua",
	"language": "lua",
	"interrupt_mode": "message",
}
```

I'm not able to easily test this b/c I'm on nixOS and ilua just doesn't work on my OS for some reason that I'm unable to figure out quickly.

You should be able to put that json into a file `~/.local/share/jupyter/kernels/ilua/kernel.json` and then molten will hopefully be able to pick it up, and fingers crossed it will actually work.

**Reply by Or1g3n** [reply]
Sweet! I'll give this a shot when I get a chance. 🤞

**Reply by Or1g3n** [reply, marked as answer]
Lord have mercy it worked!

## Steps

1. navigated to jupyter/kernels, which for me was here since I used a virtual environment:
```~/.venv/share/jupyter/kernels/```
2. duplicated the lua folder and named it nlua
![image](https://github.com/user-attachments/assets/51453136-4a81-492f-ba89-164309d2cd00)
3. within nlua folder, updated the ```kernel.json``` file:
![image](https://github.com/user-attachments/assets/c591be0b-9882-4c5b-ad03-43db5db0dc5b)
4. tested molten inside a test.lua file and viola! 
![image](https://github.com/user-attachments/assets/5251eed1-03e0-49e6-b6c5-33ede67fe918)
![image](https://github.com/user-attachments/assets/0ccf11ff-4dbd-44a1-99ab-1ac129f8137d)

**Reply by benlubas** [reply]
awesome! I'm glad that worked. I'm now super curious if we could somehow get it to use the current running neovim as the lua runtime... but I think that would require a lot more work that I probably won't ever get around to lol.

Thanks for providing detailed step by step instructions as well!

**Reply by Or1g3n** [reply]
Absolutely! Thanks a lot for the help/guidance. After playing around with this for a bit, I was thinking the same thing lol...now how do I get it connected to current session... if I ever figure it out will post back!
